independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java

```diff
@@ -245,17 +246,18 @@ void init() {
                 if (declaresObserver.contains(bean)) {
                     continue test;
                 }
-                // Declares a producer - see also second pass
-                if (declaresProducer.contains(bean)) {
-                    continue test;
-                }
                 // Instance<Foo>
                 for (InjectionPointInfo injectionPoint : instanceInjectionPoints) {
                     if (Beans.hasQualifiers(bean, injectionPoint.getRequiredQualifiers()) && Beans.matchesType(bean,
                             injectionPoint.getRequiredType().asParameterizedType().arguments().get(0))) {
                         continue test;
                     }
                 }
+                // Declares a producer - see also second pass
```

Comment:
Is there a specific reason why this is now processed after `instanceInjectionPoints`?

Reply:
There is. If `Foo`, a bean that declares a producer, is injected via `Instance<Foo>` we shall not  remove it either.

Reply:
In that case, shouldn't we have a test for this as well? ;-)

Reply:
Feel free to add one. I think the test I've added is enough ;-).

Reply:
Done